acceptance: TestDockerCLI_test_error_hints failed

acceptance.TestDockerCLI_test_error_hints [failed](https://mesolite.cluster.engflow.com/invocations/default/26c9203f-7163-464d-bcf9-9cb42ea7c1c1?testReportRun=1&testReportShard=6&testReportAttempt=1#targets-Ly9wa2cvYWNjZXB0YW5jZTphY2NlcHRhbmNlX3Rlc3Q=) on master @ [f483ecbf5c50c9c5bdc7d1f14a7ecebeb051adca](https://github.com/cockroachdb/cockroach/commits/f483ecbf5c50c9c5bdc7d1f14a7ecebeb051adca):


```
=== RUN   TestDockerCLI_test_error_hints
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestDockerCLI_test_error_hints3438524894
    test_log_scope.go:76: use -show-logs to present logs inline
    generated_cli_test.go:196: -- test log scope end --
--- FAIL: TestDockerCLI_test_error_hints (62.92s)
=== RUN   TestDockerCLI_test_error_hints/test_error_hints.tcl
    util_docker.go:256: non-zero exit code: 1
--- FAIL: TestDockerCLI_test_error_hints/test_error_hints.tcl (53.91s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_error_hints.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50877

acceptance.TestDockerCLI_test_error_hints [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/19950897?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/19950897?buildTab=artifacts#/) on master @ [67cb06a1b702789469ccbccf8368d89790e7db03](https://github.com/cockroachdb/cockroach/commits/67cb06a1b702789469ccbccf8368d89790e7db03):


```
=== RUN   TestDockerCLI_test_error_hints
    test_log_scope.go:165: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/534f11b5b7b50f31261cf1b2945268b5/logTestDockerCLI_test_error_hints472666971
    test_log_scope.go:76: use -show-logs to present logs inline
    generated_cli_test.go:196: -- test log scope end --
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/534f11b5b7b50f31261cf1b2945268b5/logTestDockerCLI_test_error_hints472666971
--- FAIL: TestDockerCLI_test_error_hints (71.28s)
=== RUN   TestDockerCLI_test_error_hints/test_error_hints.tcl
    util_docker.go:256: non-zero exit code: 1
--- FAIL: TestDockerCLI_test_error_hints/test_error_hints.tcl (61.34s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_error_hints.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

acceptance.TestDockerCLI_test_error_hints [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/20129557?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/20129557?buildTab=artifacts#/) on master @ [5128b4772335eb67ba0fb2e6395d81cb4b815d68](https://github.com/cockroachdb/cockroach/commits/5128b4772335eb67ba0fb2e6395d81cb4b815d68):


```
=== RUN   TestDockerCLI_test_error_hints
    test_log_scope.go:165: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/534f11b5b7b50f31261cf1b2945268b5/logTestDockerCLI_test_error_hints249006735
    test_log_scope.go:76: use -show-logs to present logs inline
    generated_cli_test.go:196: -- test log scope end --
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/534f11b5b7b50f31261cf1b2945268b5/logTestDockerCLI_test_error_hints249006735
--- FAIL: TestDockerCLI_test_error_hints (73.31s)
=== RUN   TestDockerCLI_test_error_hints/test_error_hints.tcl
    util_docker.go:256: non-zero exit code: 1
--- FAIL: TestDockerCLI_test_error_hints/test_error_hints.tcl (62.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_error_hints.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/cli/interactive_tests/common.tcl
+++ b/pkg/cli/interactive_tests/common.tcl
@@ -73,7 +73,7 @@ proc eexpect {text} {
 }
 
 # eexpect_re is like eexpect, but takes a regular expression argument
-# instead of a text string
+# instead of a text string.
 proc eexpect_re {text} {
   expect {
     -re $text {}
@@ -86,6 +86,44 @@ proc interrupt {} {
     report "INTERRUPT TO FOREGROUND PROCESS"
     send "\003"
     sleep 0.4
+}
+
+# Convenience function that sends Ctrl+C and waits more reliably for process to exit.
+# This uses a retry approach similar to stop_server but adapted for expect environment.
+proc interrupt_and_wait {} {
+    report "INTERRUPT AND WAIT FOR FOREGROUND PROCESS"
+
+    # Send the interrupt signal
+    send "\003"
+
+    # Use a retry loop similar to stop_server's wait mechanism.
+    # Try up to 30 times with 1 second intervals, like stop_server does.
+    try {
+        # Set shorter timeout for the retry loop
+        set timeout 1
+
+        for {set i 1} {$i <= 30} {incr i} {
+            # Try to expect the shell prompt with a short timeout (1 second per attempt).
+            expect {
+                ":/# " {
+                    report "FOREGROUND PROCESS EXITED SUCCESSFULLY (after $i attempt(s))"
+                    return
+                }
+                timeout {
+                    # Process might still be running, continue waiting.
+                    report "still waiting for process to exit (attempt $i/30)"
+                    # The sleep is implicit in the timeout, so we don't need an additional sleep here.
+                }
+            }
+        }
+
+        # If we get here, the process didn't exit cleanly within 30 seconds.
+        report "TIMEOUT: foreground process still running after 30 attempts"
+        error "Failed to interrupt foreground process - it did not exit within 30 seconds"
+    } finally {
+        # Restore original timeout.
+        set timeout 45
+    }
 }
 
 # Convenience function that sends Ctrl+D to the monitored process.

--- a/pkg/cli/interactive_tests/test_error_hints.tcl
+++ b/pkg/cli/interactive_tests/test_error_hints.tcl
@@ -69,9 +69,7 @@ end_test
 # server.
 
 set spawn_id $shell_spawn_id
-interrupt
-interrupt
-eexpect ":/# "
+interrupt_and_wait
 
 send "$argv start-single-node --listen-addr=localhost --certs-dir=$certs_dir --store=logs/db\r"
 eexpect "restarted pre-existing node"
@@ -90,13 +88,10 @@ eexpect "ERROR: node is running secure mode, SSL connection required"
 eexpect ":/# "
 end_test
 
-
 # Check what happens when attempting to connect to something
 # that is not a CockroachDB server.
 set spawn_id $shell_spawn_id
-interrupt
-interrupt
-eexpect ":/# "
+interrupt_and_wait
 
 start_test "Connecting an insecure RPC client to a non-CockroachDB server"
 # In one shell, start a bogus server
@@ -110,8 +105,7 @@ eexpect "insecure\r\n"
 set spawn_id $shell_spawn_id
 eexpect "connected"
 eexpect "PRI * HTTP/2.0"
-interrupt
-eexpect ":/# "
+interrupt_and_wait
 # Check that cockroach node drain becomes suitably confused.
 set spawn_id $client_spawn_id
 eexpect "ERROR: server closed the connection."
@@ -133,8 +127,7 @@ eexpect "insecure\r\n"
 set spawn_id $shell_spawn_id
 eexpect "connected"
 eexpect "cockroach sql"
-interrupt
-eexpect ":/# "
+interrupt_and_wait
 # Check that cockroach sql becomes suitably confused.
 set spawn_id $client_spawn_id
 eexpect "ERROR: server closed the connection."

